BAH-435 Date picker date in appointment list/calendar view

smartyrad opened the pull request:
Date picker date in appointment list/calendar view can be cleared with the delete button

smartyrad commented:
@mks-d @djazayeri @shruthidipali Can you please review this PR?
 I am not used to writing tests with Jasmine and hence any help with that will be appreciated :)

sumanmaity112 commented:
Hi @smartyrad, 

Can you squash two commits into one? Also, fix the codeclimate issues

smartyrad commented:
@sumanmaity112 Sorry for the delay, I have squashed the commits and have also fixed the issues pointed out by codeclimate. I would like to get your help in creating the tests for the same.

mks-d commented on ui/app/appointments/directives/datePicker.js:
```diff
@@ -22,15 +22,23 @@ angular.module('bahmni.appointments')
-            $scope.$watch("viewDate", $scope.onChange($scope.viewDate));
+            $scope.$watch("viewDate", function (viewDate) {
+                if (!viewDate) {
+                    $scope.viewDate = $scope.lastValidDate;
```
Is there never a case where `$scope.lastValidDate` is not yet defined here?

Isn't it necessary to at least initialize `$scope.lastValidDate` as today?

mks-d commented on ui/test/unit/appointments/directives/datePicker.spec.js:
```diff
@@ -23,7 +23,7 @@ describe('DatePicker', function () {
+    it('should init view date to the last valid date if undefined', function () {
         scope.viewDate = undefined;
         createElement();
         expect(scope.viewDate).toEqual(moment().startOf('day').toDate());
```
Shouldn't this check that `scope.viewDate` equals `scope.lastValidDate`?

smartyrad commented:
@mks-d How shall I go about fixing the Travis build? I have not updated this PR with the latest code yet but the build failures are exactly the same. The main issue is I'm not yet very sure of how tests run using Jasmine.

mks-d commented:
@smartyrad what is the minimal set of changes to highlight the Jasmine tests problem (without in fact impacting the app at runtime)? Just changing the binding for `onChange` from `"&"` to `"="`?

smartyrad commented:
Hi @mks-d , Yes, just changing the binding leads to a failure

sumanmaity112 commented:
Hi @mks-d, @smartyrad 
I think it should be fine.

mks-d commented:
Thanks @sumanmaity112 .
So @smartyrad you should be able to move forward with this now.

smartyrad commented:
@mks-d Thanks a ton! I'll get to work right away

mks-d commented:
@smartyrad you don't necessarily have to close this PR, you can keep submitting incremental commits.

smartyrad commented:
@mks-d  Actually, I had created a new branch locally. So now instead of creating a pr from my master branch, I have a branch dedicated for this issue. That's the reason a new pr is automatically created and I had to close this one to avoid duplication. I'll be fixing the failures in the other pr and will ping you as soon as I'm done :)
